File: 백준/Gold/1987. 알파벳/알파벳.py
# 1987 알파벳
R, C = list(map(int, input().split(" ")))
graph = [list(map(lambda x: ord(x)-ord("A"), input())) for _ in range(R)]
delta_list = [(1, 0), (-1, 0), (0, 1), (0, -1)]


def iterative_dfs():
    visited_alphabet = (1 << 26) - 1
    visited_bitmask = (1 << 26) - 1
    pos_bitmask = (1 << 5) - 1
    # C CCCC CCCC XXXX XYYY YYVV VVVV VVVV VVVV VVVV VVVV VVVV
    stack = [visited_alphabet]
    max_count = 0

    while stack:
        state = stack.pop()
        available_alphabet = state & visited_bitmask
        pos_y = (state >> 26) & pos_bitmask
        pos_x = (state >> 31) & pos_bitmask
        count = state >> 36
        alphabet_order = graph[pos_y][pos_x]
        
        if available_alphabet & (1 << alphabet_order):
            available_alphabet ^= (1 << alphabet_order)
            max_count = max(max_count, count + 1)
            
            for dx, dy in delta_list:
                new_x, new_y = pos_x + dx, pos_y + dy
                if 0 <= new_x < C and 0 <= new_y < R:
                    stack.append(((count + 1) << 36) + (new_x << 31) + (new_y << 26) + available_alphabet)
            
            available_alphabet ^= (1 << alphabet_order)

    return max_count

answer = iterative_dfs()
print(answer)

File: 백준/Gold/1987. 알파벳/알파벳.cs
namespace ProblemSolving;

public class Boj01987
{
    public static void Main(string[] args)
    {
        var _ = Console.ReadLine()
                       .Split(' ')
                       .Select(int.Parse)
                       .ToArray();
        var (r, c) = (_[0], _[1]);
        var graph = new int[r, c];
        for (var i = 0; i < r; i++)
        {
            var s = Console.ReadLine();
            for (var j = 0; j < c; j++)
            {
                graph[i, j] = s[j] - 'A';
            }
        }

        var answer = 0L;
        // Bit arrangement: _______C CCCCCCCC ___XXXXX ___YYYYYY ______VV VVVVVVVV VVVVVVVV VVVVVVVV
        var deltaList = new List<(int, int)> { (1, 0), (-1, 0), (0, 1), (0, -1) };
        var stack = new Stack<long>();
        stack.Push((1 << 26) - 1);
        while (stack.Count > 0)
        {
            var state = stack.Pop();
            var availableAlphabet = (int)state;
            var y = state >> 32 & 0xff;
            var x = state >> 40 & 0xff;
            var count = state >> 48;
            var alphabet = graph[y, x];
            if ((availableAlphabet & (1 << alphabet)) == 0) continue;
            availableAlphabet ^= (1 << alphabet);
            answer = Math.Max(answer, count + 1);
            foreach (var (dx, dy) in deltaList)
            {
                var nx = x + dx;
                var ny = y + dy;
                if (nx < 0 || nx >= c || ny < 0 || ny >= r) continue;
                var nextState = count + 1 << 48 | nx << 40 | ny << 32 | (uint)availableAlphabet;
                stack.Push(nextState);
            }

            availableAlphabet ^= (1 << alphabet);
        }

        Console.WriteLine(answer);
    }
}